imap_processing/idex/idex_packet_parser.py

```diff
@@ -0,0 +1,286 @@
+    it exists on the local file system.
+
+    Parameters:
+        filename (str):  The path to the CDF file to read
+
+     Example:
+        >>> # Print out the data in a single packet
+        >>> from imap_processing.idex.idex_packet_parser import IDEXPacketParser
+        >>> l0_file = "imap_processing/idex/tests/imap_idex_l0_20230725_v01-00.pkts"
+        >>> l0_data = IDEXPacketParser(l0_file)
+        >>> print(l0_data.epochs)
+        >>> print(l0_data.data)
+    """
+
+    def __init__(self, idex_packet_file: str):
+        """
+        This function takes in a local l0 pkts file and performs all of the decom work
+        directly in __init__().
+
+        Parameters:
+            idex_packet_file (str):  The path and filename to the L0 file to read
+
+        Returns:
+            Populates {object}.data, where data is an xarray DataSet
+
+        Notes:
+            Currently assumes one L0 file will generate exactly one l1a file
+
+            There are a few things that must be determined in the future:
+            * high_sample_rate and low_sample_rate must be multiplied by
+              some scale factor that will be decided upon in the future
+            * IDEX_Trigger is a placeholder variable for now
+        """
+
+        idex_xtce = f"{packet_definition_directory}/idex_packet_definition.xml"
+        idex_definition = xtcedef.XtcePacketDefinition(xtce_document=idex_xtce)
+        idex_parser = parser.PacketParser(idex_definition)
+
+        idex_binary_data = bitstring.ConstBitStream(filename=idex_packet_file)
+        idex_packet_generator = idex_parser.generator(idex_binary_data)
+
+        self.coords = {}
+        self.data = {}
+        self.dust_events = {}
+        self.scitype_to_names = {
+            2: "TOF_High",
+            4: "TOF_Low",
+            8: "TOF_Mid",
+            16: "Target_Low",
+            32: "Target_High",
+            64: "Ion_Grid",
+        }
+
+        for packet in idex_packet_generator:
+            if "IDX__SCI0TYPE" in packet.data:
+                scitype = packet.data["IDX__SCI0TYPE"].raw_value
+                event_number = packet.data["IDX__SCI0EVTNUM"].derived_value
+                if scitype == 1:
+                    time_of_impact = (
+                        packet.data["SHCOARSE"].derived_value
+                        + TWENTY_MICROSECONDS * packet.data["SHFINE"].derived_value
+                    )
+                    self._log_packet_info(
+                        time_of_impact, packet
+                    )  # These are for our logs
+                    self.dust_events[event_number] = IDEXRawDustEvent(time_of_impact)
+                if (
+                    scitype in self.scitype_to_names
+                ):  # Populate the IDEXRawDustEvent with 1's and 0's
+                    raw_science_data = packet.data["IDX__SCI0RAW"].raw_value
+                    self.dust_events[event_number].append_raw_data(
```

Comment:
Do you want to pass the packet in again and do the logic within this method? `append_raw_data(packet)`

Also, does append matter on the order? What happens if these come in out of order and therefore your appends put science type 16 before science type 8? Do we need checks for that?

Reply:
I believe when receive the L0 data, the packets we receive should be time ordered